File: custom_app/public/js/bom_tree_view.js
// custom_bom_tree_view.js

frappe.ui.form.on('BOM Creator', {
    refresh: function (frm) {
        if (frm.doc.docstatus == 1) {
            frappe.call({
                method: "custom_app.public.py.bom_tree_view.execute",
                args: {
                    bom_creator_name: frm.doc.name
                },
                callback: function (r) {
                    if (r.message) {
                        let data = r.message;
                        console.log("data", data)
                        render_tree_view(data, frm);
                    }
                }
            });
        }
        frappe.call({
            method: "custom_app.public.py.bom_tree_view.calculate_total_amount",
            args: {
                name: frm.doc.name
            }
        })
    },

});


function render_tree_view(data, frm) {
    let tree_view_html = '';

    let nodes = {};
    let tree = [];

    data.forEach(row => {
        if (row.operation_type == "Parent") {
            let node = {
                operation_type: row.operation_type,
                operation_name: row.operation_name,
                children: []
            }

            nodes[row.name] = node;
            tree.push(node);
        }
        else if (row.operation_type == "Finished Item Operation") {
            let parentNode = nodes[row.name];
            if (parentNode) {
                parentNode.children.push({
                    operation_name: row.operation_name,
                    operation_type: row.operation_type,
                });
            }
        }
        else if (row.operation_type == "Sub Assembly Item") {
            let parentNode = nodes[row.name];
            if (parentNode) {
                var subAssemblyNode = {
                    item_code: row.item_code,
                    operation_name: row.operation_name,
                    operation_type: row.operation_type,
                    children: []
                };
                parentNode.children.push(subAssemblyNode);
                // Also add the subAssemblyNode to nodes so that it can be a parent to further nodes
                nodes[row.operation_name] = subAssemblyNode;
            }
        }
        else if (row.operation_type === "Sub Assembly Operation") {
            let data1 = nodes[row.name]
            data1.children.forEach(row1 => {
                if (row1.operation_type == "Sub Assembly Item") {
                    let subAssembly = row1
                    if (subAssembly) {
                        subAssembly.children.push({
                            item_code: row.item_code,
                            operation_name: row.operation_name,
                            operation_type: row.operation_type,
                        });
                    }
                }
            })
        }
    })

    function renderNode(node) {
        let html = `<li>${node.operation_name}`;
        if (node.children && node.children.length > 0) {
            html += `<ul>`;
            node.children.forEach(child => {
                html += renderNode(child);  // Recursively render child nodes
            });
            html += `</ul>`;
        }
        html += `</li>`;
        return html;
    }

    // Render only the main parent and its direct children
    tree.forEach(rootNode => {
        tree_view_html += `<ul>${renderNode(rootNode)}</ul>`;
    });

    // frm.fields_dict['custom_tree_view'].html(tree_view_html);
}

File: custom_app/public/py/bom_tree_view.py
import frappe
from frappe import _

@frappe.whitelist()
def execute(bom_creator_name=None):
    data = [] 

    # Fetch the BOM data
    bom_data = frappe.db.sql("""
        SELECT 
            bc.name as bom_creator,
            bc.item_code as finished_item
        FROM 
            `tabBOM Creator` bc
        WHERE 
            bc.docstatus = 1 and
            bc.name = %s
        ORDER BY 
            bc.name
    """,(bom_creator_name), as_dict=True)

    # print("\n\nbom_data", bom_data)

    # Fetch sub-assembly items
    sub_assembly_items = frappe.db.sql("""
        SELECT 
            bci.parent as bom_creator, 
            bci.item_code as sub_assembly_item
        FROM 
            `tabBOM Creator Item` bci
        WHERE 
            bci.is_expandable = 1 and
            bci.parent = %s
        ORDER BY 
            bci.parent
    """,(bom_creator_name), as_dict=True)

    # print("\nsub_assembly_items", sub_assembly_items)

    # Fetch finished item operations
    finished_operations = frappe.db.sql("""
        SELECT 
            ob.parent as bom_creator, 
            ob.operation
        FROM 
            `tabOperastion Bom` ob
        WHERE 
            ob.parent =%s
    """,(bom_creator_name), as_dict=True)

    # print("\nfinished_operations", finished_operations)

    # Fetch sub-assembly operations
    sub_assembly_operations = frappe.db.sql("""
        SELECT 
            sao.parent as bom_creator, 
            sao.item_code as sub_assembly_item,
            sao.operation
        FROM 
            `tabSub Assemblies Oprastion` sao
        WHERE 
            sao.parent =%s
    """,(bom_creator_name), as_dict=True)

    # print("\nsub_assembly_operations", sub_assembly_operations)

    # Process the data into a tree structure
    for row in bom_data:
        bom_data1 = {
            "name": row.bom_creator,
            "operation_type": "Parent",
            "operation_name": row.finished_item,
        }

    for op in finished_operations:
        if op.bom_creator == row.bom_creator:
            finished_operations1 = {
                "name": row.bom_creator,
                "operation_type": "Finished Item Operation",
                "operation_name": op.operation
            }

    for sub_item in sub_assembly_items:        
        if sub_item.bom_creator == row.bom_creator:
            sub_assembly_items1 = {
                "name": sub_item.bom_creator,
                "item_code": sub_item.sub_assembly_item,
                "operation_type": "Sub Assembly Item",
                "operation_name": sub_item.sub_assembly_item,
            }
    for op in sub_assembly_operations:
        if op.bom_creator == row.bom_creator and op.sub_assembly_item == sub_item.sub_assembly_item:
            sub_assembly_operations1 = {
                "name": sub_item.bom_creator,
                "item_code": sub_item.sub_assembly_item,
                "operation_type": "Sub Assembly Operation",
                "operation_name": op.operation
            }

    
    
    return bom_data1, finished_operations1, sub_assembly_items1, sub_assembly_operations1



@frappe.whitelist()
def calculate_total_amount(name):
    self = frappe.get_doc("BOM Creator", name)

    main_operation = self.custom_operastion_bom
    items = self.items
    sub_assemblies_oprastion = self.custom_sub_assemblies_oprastion

    operation_amount = 0
    sub_assemblies_rawmaterial_amount = 0
    for item in main_operation:
        operation_amount += item.amount
    for sub_operation in sub_assemblies_oprastion:
        operation_amount += sub_operation.amount
    for items in items:
        sub_assemblies_rawmaterial_amount += items.amount

    total_amount = operation_amount + sub_assemblies_rawmaterial_amount
    frappe.db.set_value("BOM Creator", name, "custom_sub_assemblies__raw_materials_amount", sub_assemblies_rawmaterial_amount)
    frappe.db.set_value("BOM Creator", name, "custom_operation_amount", operation_amount)
    frappe.db.set_value("BOM Creator", name, "custom_total_amount", total_amount)